x-pack/legacy/plugins/ml/public/util/custom_url_utils.ts

```diff
@@ -0,0 +1,206 @@
+ */
+function escapeForKQL(value: string): string {
+  return value.replace(/\"/g, '\\"');
+}
+
+// Builds a Kibana dashboard or Discover URL from the supplied config, with any
+// dollar delimited tokens substituted from the supplied anomaly record.
+function buildKibanaUrl(urlConfig: UrlConfig, record: AnomalyRecordSource) {
+  const urlValue = urlConfig.url_value;
+  const URL_LENGTH_LIMIT = 2000;
+
+  const isLuceneQuery = urlValue.includes('language:lucene');
+  const queryLanguageEscapeCallback = isLuceneQuery ? escapeForElasticsearchQuery : escapeForKQL;
+
+  type GetResultTokenValue = (v: string) => string;
+
+  // Compose callback for characters escaping and encoding.
+  const getResultTokenValue: GetResultTokenValue = flow(
+    queryLanguageEscapeCallback,
+    // Kibana URLs used rison encoding, so escape with ! any ! or ' characters
+    (v: string): string => v.replace(/[!']/g, '!$&'),
+    encodeURIComponent
+  );
+
+  return String(urlValue)
+    .replace('$earliest$', record.earliest)
+    .replace('$latest$', record.latest)
+    .replace(
+      /(.+query:')([^']*)('.+)/,
+      (fullMatch, prefix: string, queryString: string, postfix: string) => {
+        const [resultPrefix, resultPostfix] = [prefix, postfix].map(str =>
+          str.replace(/\$(\w+)\$/g, (match, name: string) => {
+            // Use lodash get to allow nested JSON fields to be retrieved.
+            const tokenValue: string | undefined = get(record, name);
+            // If property not found string is not replaced.
+            return tokenValue === undefined ? match : getResultTokenValue(tokenValue);
+          })
+        );
+
+        let availableCharactersLeft = URL_LENGTH_LIMIT - resultPrefix.length - resultPostfix.length;
+        const queryFields = queryString
+          // Split query string by AND operator.
+          .split(/\sand\s/i)
+          // Get property name from `influencerField:$influencerField$` string.
+          .map(v => v.split(':')[0]);
+
+        const queryParts: string[] = [];
+        const joinOperator = ' AND ';
+
+        for (let i = 0; i < queryFields.length; i++) {
+          const field = queryFields[i];
+          // Use lodash get to allow nested JSON fields to be retrieved.
+          const tokenValues: string[] = get(record, field) || null;
+          if (tokenValues === null) {
+            continue;
+          }
+          // Create a pair `influencerField:value`.
+          // In cases where there are multiple influencer field values for an anomaly
+          // combine values with OR operator e.g. `(influencerField:value or influencerField:another_value)`.
+          let result = tokenValues
+            .map(value => `${field}:"${getResultTokenValue(value)}"`)
+            .join(' OR ');
+          result = tokenValues.length > 1 ? `(${result})` : result;
+
+          availableCharactersLeft -= result.length - (i === 0 ? 0 : joinOperator.length);
```

Comment:
i think it would be worth adding comments to explain why we're keeping track of the character count. That we're trying to build up a string that's under the allowed length but that isn't corrupted due to it missed its end.